posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,196 @@
+```python
+from scipy.ndimage import laplace
+import numpy as np
+import cupy as cp
+
+numpy_img = np.random.normal(size=(256, 256))
+cupy_img = cp.array(numpy_img)
+
+print(type(laplace(numpy_img)))
+# <class 'numpy.ndarray'>
+
+print(type(laplace(cupy_img)))
+# <class 'cupy._core.core.ndarray'>
+```
+
+Since Python doesn't provide standard tools for using multiple dispatch there are several dedicated libraries:
+* _multimethod_ - pure Python implementation of multiple dispatch with caching of argument types,
+* _plum-dispatch_ - implementation of multiple dispatch that follows the ideas from Julia,
+* _uarray_ is a generic backend/multiple dispatch library developed to provide NumPy-independent but similar to NumPy's dispatch functionality.
+
+First two options are very similar in functionality and usability. But the last option, in addition to multiple dispatch, offers granular control using context managers and provides a way to switch backends for the same array type. It's also possible to make the `uarray` dispatcher work without a context manager making the mechanism implicit and similar to other dispatch libraries. Whether we should encourage the implicit registration of backends or not is an open design issue and there's previous discussion on the topic in [SciPy #14266](https://github.com/scipy/scipy/issues/14266). Hameer Abbasi, one of the `uarray` authors, wrote [a blog post](https://labs.quansight.org/blog/2019/07/uarray-update-api-changes-overhead-and-comparison-to-__array_function__/) about the motivation for the `uarray` and how it compares to NumPy's `__array_function__` dispatch mechanism.
+
+## It's already happening
+
+SciPy has GPU and distributed arrays support [on the roadmap](http://scipy.github.io/devdocs/dev/roadmap.html), but adding this support directly to the library would inflate the scope and increase the maintenance burden significantly. scikit-learn has [explicitly stated](https://scikit-learn.org/stable/faq.html#will-you-add-gpu-support) that they cannot afford GPU-specific code in their codebase. A dispatching system is a low-maintenance solution for extending the functionality of the API.
+
+In 2019 Peter Bell added support for backend switching to `scipy.fft` module ([PR #10383](https://github.com/scipy/scipy/pull/10383)) using `uarray`. Now it's possible to tell SciPy to use, for example, CuPy backend, for computing FFT when CuPy's array is passed to functions from `scipy.fft`.
+
+In this recent blog post, Anirudh Dagar explores array interoperability protocols: `__array_function__` (NEP 18), `__array_namespace__` (NEP 47) and `uarray`. The blog post includes a [section comparing the Array API and `uarray`](https://labs.quansight.org/blog/2021/10/array-libraries-interoperability/#protocol_differences) approaches for interoperability.
```

Comment:
```suggestion
In a recent blog post titled [Array Libraries Interoperability](https://labs.quansight.org/blog/2021/10/array-libraries-interoperability/), Anirudh Dagar explores array interoperability protocols: `__array_function__` (NEP 18), `__array_namespace__` (NEP 47) and `uarray`. The blog post includes a [section comparing the Array API and `uarray`](https://labs.quansight.org/blog/2021/10/array-libraries-interoperability/#protocol_differences) approaches for interoperability.
```
Add link and reword with the title of the blog to follow accessibility guidelines.